### PR TITLE
DBZ-7167 Provide a public API from the connector implementations to retrieve the list of matching collections or tables based on the different include-/exclude lists

### PR DESCRIPTION
**Pull request by harveyyue**
https://issues.redhat.com/browse/DBZ-7167
Will implement the connector interface method `getMatchingCollections` to rest of connectors

### CODE REVIEW

**Comment by jpechane** [comment]
@harveyyue Thanks a lot for the PR. The intend of this change is twofold - the migration of the code **and** using the same code in the REST extension and snapshot/streaming interface. So there should be an overlap with changes on `io.debezium.relational.RelationalSnapshotChangeEventSource.getAllTableIds(RelationalSnapshotContext<P, O>)` and `io.debezium.pipeline.source.AbstractSnapshotChangeEventSource.determineDataCollectionsToBeSnapshotted(Collection<T>, Set<Pattern>)`. We want to make sure that the same pice of code is used both in UI provided metadata and in the connector itself to make sure the operate on the same set.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/rest/DebeziumMongoDbConnectorResource.java
Comment (rk3rn3r) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/rest/DebeziumMongoDbConnectorResource.java:51]: @jpechane Would it maybe make sense to add another interface on the connector classes/implementations and let `getConnector()` return this one here? That way we could move this code here into the FilterValidationResource class instead of having this duplicate coder everywhere and an unnecessary cast. Wdyt?

Comment (jpechane) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/rest/DebeziumMongoDbConnectorResource.java:51]: Is the typecasting even necessary? IMHO it could be pulled-up to the parent class and just call `return getConnector().getMatchingCollections(configuration);`

Comment (jpechane) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/rest/DebeziumMongoDbConnectorResource.java:51]: ok, I see what you mean. We don't have a common parent for all Debezium connectors. Maybe instead of adding a new interface we should add a common parent class for all connectors lakie we do for tasks so intorducing `BaseConnector` class?

Comment (rk3rn3r) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/rest/DebeziumMongoDbConnectorResource.java:51]: Agreed!

**Review by jpechane — COMMENTED** [review]
@harveyyue It looks very neat, thanks a lot!

I left just two monir comments related to code simplification and typing changes.

@rk3rn3r Could you please confirm there would be no issue if the resource interface would be generic and have a default method?

**Review by jpechane — APPROVED** [review]
@harveyyue LGTM, thanks a lot! Waiting for @rk3rn3r to comment before the merge.

**Comment by harveyyue** [comment]
informix pr: https://github.com/debezium/debezium-connector-informix/pull/14

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/rest/DebeziumMySqlConnectorResource.java
Comment (rk3rn3r) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/rest/DebeziumMySqlConnectorResource.java]: I don't think so (if I got you right) because `getConnector()` and `getMatchingCollections()` have entirely connector specific code. There is not much in common or else we have to add a bunch of other methods to the interface to provide the diverging `*ConnectorConfig` instances `*Connection` classes etc.

File: debezium-core/src/main/java/io/debezium/rest/ConnectionValidationResource.java
Comment (rk3rn3r) [reply debezium-core/src/main/java/io/debezium/rest/ConnectionValidationResource.java]: Ah, now I see what you meant. :bulb:

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/rest/DebeziumMySqlConnectorResource.java
Comment (rk3rn3r) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/rest/DebeziumMySqlConnectorResource.java]: never mind, I think it looks very good now!

**Comment by rk3rn3r** [comment]
LGTM. It looks like after marking some other tests as flaky now others (maybe some that did not run before because of earlier tests failing which now get skipped) are failing. I re-triggered builds for failing jobs and see how they resolve. I will merge the PR after this test run finished.
Thx a lot for a very important and valuable contribution @harveyyue!

File: debezium-core/src/main/java/io/debezium/connector/common/BaseSourceConnector.java
Comment (rk3rn3r) [inline debezium-core/src/main/java/io/debezium/connector/common/BaseSourceConnector.java:25]: @jpechane  I also don't think we should use a response model from the REST extension here `List<DataCollection>`? Would TableId or something else match better the "internal connector" use-case while for the UI we then need to translate that into the DataCollection model?

**Comment by harveyyue** [comment]
@rk3rn3r Have implemented db2, do you mean the other cassandra and spanner connectors?

File: debezium-core/src/main/java/io/debezium/connector/common/BaseSourceConnector.java
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/connector/common/BaseSourceConnector.java:25]: @rk3rn3r Well, you can't as MongoDB does not use `TableId`. What you can do it to have generic parameter `T extends DataCollectionId`.

Comment (rk3rn3r) [reply debezium-core/src/main/java/io/debezium/connector/common/BaseSourceConnector.java:25]: I was not saying it must be `TableId`, just a suggestion (and I think those are new methods and if we decide that the method return `TableId` we could use a meaningful mapping to the TableId fields, this is required anyway). We could also create a value object as a copy from the `DataCollection` REST resonse model but I don't think it's a good idea to use the same entity for the connectors and for the REST response, although they are related there might be reasons in the future to diverge on the connector side implementation but we like to keep the REST response model unchanged.

**Comment by rk3rn3r** [comment]
@harveyyue No, please have the look at the CI test error for Db2:

```
 Error:  COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
Error:  /home/runner/work/debezium/debezium/db2/src/main/java/io/debezium/connector/db2/Db2Connector.java:[33,8] io.debezium.connector.db2.Db2Connector is not abstract and does not override abstract method getMatchingCollections(io.debezium.config.Configuration) in io.debezium.connector.common.BaseSourceConnector
```
https://github.com/debezium/debezium/actions/runs/7139864033/job/19557779538?pr=5073#step:7:130

**Comment by harveyyue** [comment]
> @harveyyue No, please have the look at the CI test error for Db2:
> 
> ```
>  Error:  COMPILATION ERROR : 
> [INFO] -------------------------------------------------------------
> Error:  /home/runner/work/debezium/debezium/db2/src/main/java/io/debezium/connector/db2/Db2Connector.java:[33,8] io.debezium.connector.db2.Db2Connector is not abstract and does not override abstract method getMatchingCollections(io.debezium.config.Configuration) in io.debezium.connector.common.BaseSourceConnector
> ```
> 
> https://github.com/debezium/debezium/actions/runs/7139864033/job/19557779538?pr=5073#step:7:130

Yes, there is another pr for db2 connector:
https://github.com/debezium/debezium-connector-db2/pull/124

**Comment by jpechane** [comment]
@rk3rn3r It is a different repository, it must be different PR

**Comment by rk3rn3r** [comment]
@harveyyue never mind. I am not 100% in shape today and in my main IDE workspace I have db2 repo symlinked. what make me think it is in the main repo. sorry :see_no_evil:

**Comment by rk3rn3r** [comment]
But it still leaves the question what other entity we can use as value object to be returned by `List<Something> getMatchingCollections(Configuration config)`. I don't think it should be `io.debezium.rest.model.DataCollection` as this is purely meant as REST response model. @jpechane

**Comment by jpechane** [comment]
Well, and here is another cofusion. We have `io.debezium.spi.schema.DataCollectionId` and IMHO it should be returned. I am sorry for not noticing the package name.

**Comment by harveyyue** [comment]
> Well, and here is another cofusion. We have `io.debezium.spi.schema.DataCollectionId` and IMHO it should be returned. I am sorry for not noticing the package name. So the return value will be `T extends DataCollectionId` so releational connectors will return TableId and MongoDB will return CollectionId.

@rk3rn3r @jpechane Done.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlSnapshotChangeEventSource.java
Comment (rk3rn3r) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlSnapshotChangeEventSource.java:253]: Shouldn't similar code above ( https://github.com/debezium/debezium/pull/5073/files#diff-63840528f25c3335e78d8b91c2ea67602750af012c1620d9981f11cb476b480aR144-R154 )  not also use the new filter code? Maybe it must even be merged into the new method to handle only "readable databases". Whatever this is....

**Comment by rk3rn3r** [comment]
LGTM @jpechane after those changes got applied. We could easily file Jiras for my other comment/s. wdyt?
I re-triggered tests to see if failures were intermittent or constant errors.

**Comment by jpechane** [comment]
@harveyyue LGTM, applied, thanks a lot!
@rk3rn3r Please perpare the folow-up Jira. Please make sure that the task to use the new API in snapshot code is included.

**Pull request merged**
